lib/action_subscriber/publisher/async/in_memory_adapter.rb

```diff
@@ -0,0 +1,105 @@
+require "thread"
+
+module ActionSubscriber
+  module Publisher
+    module Async
+      class InMemoryAdapter
+        attr_reader :async_queue
+
+        def initialize
+          @async_queue = AsyncQueue.new
+        end
+
+        def publish(route, payload, exchange_name, options = {})
+          message = Message.new(route, payload, exchange_name, options)
+          async_queue.push(message)
+          nil
+        end
+      end
+
+      class Message
+        attr_reader :route, :payload, :exchange_name, :options
+
+        def initialize(route, payload, exchange_name, options)
+          @route = route
+          @payload = payload
+          @exchange_name = exchange_name
+          @options = options
+        end
+      end
+
+      class UnableToPersistMessageError < ::StandardError
+      end
+
+      class AsyncQueue
+        attr_reader :consumer, :queue, :supervisor
+
+        MAXIMUM_QUEUE_SIZE = 1_000_000.freeze
+
+        if ::RUBY_PLATFORM == "java"
+          NETWORK_ERRORS = [::Java::ComRabbitmqClient::AlreadyClosedException, ::Java::JavaIo::IOException]
+        else
+          NETWORK_ERRORS = [::Bunny::Exception, ::Timeout::Error, ::IOError]
+        end
+
+        def initialize
+          @queue = ::Queue.new
+          create_and_supervise_consumer!
+        end
+
+        def push(message)
+          # TODO: How do we exert backpressure here? Do we block the thread, or do we raise an error?
+          fail UnableToPersistMessageError, "Queue is full, messages will be dropped." if queue.size > MAXIMUM_QUEUE_SIZE
+
+          queue.push(message)
+        end
+
+      private
+
+        def create_and_supervise_consumer!
+          @consumer = create_consumer
+          @supervisor = ::Thread.new do
```

Comment:
Supervisor and consumer are collected when the program terminates (run forever). Maybe a shutdown trigger is needed so it can empty its queue before terminating? Actually, there probably is a possible race condition where the request finishes and exits the process, but the publisher has not successfully delivered the message. I'm going to need to think about this.